runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -169,98 +227,189 @@ void iree_hal_cuda_tracing_context_free(
+  }
+
+  // Free an event list that was previously created. There is some book-keeping
+  // to keep tracy happy, and then we remove the elements from the
+  // passed in event_list and add them to the front of the free-list.
+
+  // If this event list has never been submitted we still need to add values to
+  // the timeline otherwise tracy will not behave correctly.
+  if (!event_list->head->next_submission) {
+    iree_hal_cuda_tracing_context_event_t* event = event_list->head;
+    while (event) {
+      uint32_t query_id = (uint32_t)(event - &context->event_pool[0]);
+      iree_tracing_gpu_zone_notify(context->id, query_id, 0);
+      event = event->next_in_command_buffer;
     }
   }
 
-  IREE_TRACE_ZONE_END(z0);
+  if (!context->event_freelist_head) {
+    context->event_freelist_head = event_list->head;
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+  event_list->head->next_submission = NULL;
+  event_list->tail->next_in_command_buffer = context->event_freelist_head;
+  context->event_freelist_head = event_list->head;
+
+  event_list->head = NULL;
+  event_list->tail = NULL;
+  iree_slim_mutex_unlock(&context->event_mutex);
 }
 
+// Grabs the next available query out of the freelist and adds it to
+// the event_list that was passed in. Also starts the recording of the
+// event.
 static uint16_t iree_hal_cuda_stream_tracing_context_insert_query(
-    iree_hal_cuda_tracing_context_t* context, CUstream stream) {
-  // Allocate an event from the pool for use by the query.
-  uint32_t query_id = context->query_head;
-  context->query_head = (context->query_head + 1) % context->query_capacity;
-
-  // TODO: check to see if the read and write heads of the ringbuffer have
-  // overlapped. If they have we could try to collect but it's not guaranteed
-  // that collection will complete (e.g. we may be reserving events for use in
-  // graphs that haven't yet been launched).
-  //
-  // For now we just allow the overlap and tracing results will be inconsistent.
-  IREE_ASSERT_NE(context->query_head, context->query_tail);
+    iree_hal_cuda_tracing_context_t* context,
+    iree_hal_cuda_tracing_context_event_list_t* event_list, CUstream stream) {
+  iree_slim_mutex_lock(&context->event_mutex);
+  IREE_ASSERT_ARGUMENT(event_list);
 
-  CUevent event = context->event_pool[query_id];
-  IREE_CUDA_IGNORE_ERROR(context->symbols, cuEventRecord(event, stream));
+  // Allocate an event from the pool for use by the query.
+  // TODO: If we have run out of our freelist, then we need to try and recover
+  // allocate events.
+  iree_hal_cuda_tracing_context_event_t* event = context->event_freelist_head;
+  context->event_freelist_head = event->next_in_command_buffer;
+  uint32_t query_id = event - &context->event_pool[0];
+  IREE_ASSERT(event->next_in_command_buffer != NULL);
+  event->next_in_command_buffer = NULL;
+
+  IREE_CUDA_IGNORE_ERROR(context->symbols, cuEventRecord(event->event, stream));
+
+  if (!event_list->head) {
```

Comment:
We have duplicated impl for this in both the stream and graph side. Worth having a utility function like `iree_hal_cuda_tracing_context_event_list_append_event` to deduplicate.